Mathlib/Data/Nat/Choose/Sum.lean

```diff
@@ -126,6 +126,21 @@ theorem sum_Icc_choose (n k : ℕ) : ∑ m ∈ Icc k n, m.choose k = (n + 1).cho
       rw [← Ico_insert_right (by omega), sum_insert (by simp),
         show Ico k (n + 1) = Icc k n by rfl, ih, choose_succ_succ' (n + 1)]
 
+/-- **Zhu Shijie's identity** aka hockey-stick identity, version with `range`.
+Summing `(i + k).choose k` for `i ∈ [0, n]` gives `(n + k + 1).choose (k + 1)`.
+
+Combinatorial interpretation: `(i + k).choose k` is the number of decompositions of `[0, i)` in
+`k + 1` (possibly empty) intervals (this follows from a stars and bars description). In particular,
+`(n + k + 1).choose (k + 1)` corresponds to decomposing `[0, n)` into `k + 2` intervals.
+By putting away the last interval (of some length `n - i`),
+we have to decompose the remaining interval `[0, i)` into `k + 1` intervals, hence the sum. -/
+lemma sum_range_add_choose (n k : ℕ) :
+    ∑ i ∈ Finset.range (n + 1), (i + k).choose k = (n + k + 1).choose (k + 1) := by
+  rw [← sum_Icc_choose, range_eq_Ico]
+  convert (sum_map _ (addRightEmbedding k) (·.choose k)).symm using 2
+  rw [map_add_right_Ico, zero_add, add_right_comm]
+  rfl
```

Comment:
What's the `rfl` doing? Is there a missing lemma?

Reply:
It's for `Icc k (n + k) = Ico k (n + k + 1)`. We have the `Set` version, in `Order.Ico_succ_right` (not imported by this file, though), but not the `Finset` version, so I'm happy to leave that to `rfl`. The other option is to port the material on set intervals from `Mathlib.Order.SuccPred.Basic`  to finset intervals, but I think it's a little bit too much for the current PR.

Reply:
Is it `Nat.Ico_succ_right`?

Reply:
Otherwise LGTM. Thanks a lot!
bors d+